File: backend/app/schemas/language_detection.py
"""Public Phase 7 language detection API contracts."""

from __future__ import annotations

from datetime import datetime
from enum import StrEnum
from typing import Literal
from uuid import UUID

from pydantic import Field, field_validator

from app.models.extraction_job import ExtractionJobStatus
from app.models.language_block_result import (
    LanguageCode,
    LanguageEligibilityReason,
    LanguageEligibilityStatus,
    LanguageSourceType,
)
from app.models.language_detection_job import (
    LanguageDetectionJobStatus,
    LanguageDetectionJobType,
)
from app.models.language_detection_run import LanguageDetectionRunStatus
from app.models.ocr_job import OCRJobStatus
from app.schemas.base import ApiSchema
from app.schemas.common import PaginationData
from app.schemas.language_internal import (
    CoverageBreakdownData,
    LanguagePresenceData,
)


class LanguageDetectionStartRequest(ApiSchema):
    document_file_id: UUID
    extraction_run_id: UUID
    ocr_run_id: UUID | None = None
    force: bool = False


class LanguageRedetectRequest(ApiSchema):
    reason: str = Field(min_length=1, max_length=1000)

    @field_validator("reason", mode="before")
    @classmethod
    def normalize_reason(cls, value: object) -> object:
        if isinstance(value, str):
            normalized = value.strip()
            if not normalized:
                raise ValueError("Reason must not be empty.")
            return normalized
        return value


class LanguageDetectionQueuedResponse(ApiSchema):
    job_id: UUID
    status: LanguageDetectionJobStatus
    progress: int = Field(ge=0, le=100)
    document_file_id: UUID
    extraction_run_id: UUID
    ocr_run_id: UUID | None = None
    reused_existing_result: bool = False
    run_id: UUID | None = None


class LanguageJobDocumentReference(ApiSchema):
    id: UUID
    base_document_code: str
    title: str
    department_id: UUID


class LanguageJobRevisionReference(ApiSchema):
    id: UUID
    revision_code: str
    full_document_code: str


class LanguageJobFileReference(ApiSchema):
    id: UUID
    filename: str
    extension: str
    sha256_hash: str


class LanguageJobRequesterReference(ApiSchema):
    id: UUID
    name: str


class LanguageDetectionDocumentStatus(StrEnum):
    """Filter values for the document inventory."""

    NOT_STARTED = "NOT_STARTED"
    QUEUED = "QUEUED"
    LOADING_CONTENT = "LOADING_CONTENT"
    DETECTING = "DETECTING"
    AGGREGATING = "AGGREGATING"
    PERSISTING = "PERSISTING"
    COMPLETED = "COMPLETED"
    PARTIALLY_COMPLETED = "PARTIALLY_COMPLETED"
    FAILED = "FAILED"
    CANCEL_REQUESTED = "CANCEL_REQUESTED"
    CANCELLED = "CANCELLED"


class LanguageDetectionDocumentListItem(ApiSchema):
    """Current file and its real extraction, OCR, and language state."""

    document: LanguageJobDocumentReference
    revision: LanguageJobRevisionReference
    file: LanguageJobFileReference
    extraction_status: ExtractionJobStatus | None = None
    ocr_status: OCRJobStatus | None = None
    language_detection_status: LanguageDetectionJobStatus | None = None
    language_progress: int | None = Field(default=None, ge=0, le=100)
    language_current_stage: str | None = None
    extraction_run_id: UUID | None = None
    ocr_run_id: UUID | None = None
    language_detection_run_id: UUID | None = None
    language_presence: LanguagePresenceData | None = None
    last_detected: datetime | None = None
    source_ready: bool


class LanguageDetectionDocumentListResponse(
    PaginationData[LanguageDetectionDocumentListItem]
):
    pass


class LanguageDetectionJobError(ApiSchema):
    code: str
    message: str


class LanguageDetectionJobListItem(ApiSchema):
    id: UUID
    document: LanguageJobDocumentReference
    revision: LanguageJobRevisionReference
    file: LanguageJobFileReference
    extraction_run_id: UUID
    ocr_run_id: UUID | None = None
    job_type: LanguageDetectionJobType
    status: LanguageDetectionJobStatus
    progress: int = Field(ge=0, le=100)
    current_stage: str | None = None
    requested_by: LanguageJobRequesterReference | None = None
    requested_at: datetime
    started_at: datetime | None = None
    completed_at: datetime | None = None
    cancelled_at: datetime | None = None
    run_id: UUID | None = None
    result_summary: dict[str, object] | None = None


class LanguageDetectionJobResponse(LanguageDetectionJobListItem):
    attempt_number: int = Field(ge=1)
    maximum_attempts: int = Field(ge=1)
    failed_at: datetime | None = None
    error: LanguageDetectionJobError | None = None


class LanguageDetectionJobListResponse(PaginationData[LanguageDetectionJobListItem]):
    pass


class LanguageDetectionCancelResponse(ApiSchema):
    id: UUID
    status: LanguageDetectionJobStatus
    progress: int = Field(ge=0, le=100)
    current_stage: str | None = None
    cancelled_at: datetime | None = None


class LanguagePresenceResponse(LanguagePresenceData):
    pass


class LanguageCoverageResponse(ApiSchema):
    block_coverage: CoverageBreakdownData
    character_coverage: CoverageBreakdownData
    preliminary: bool = True
    disclaimer: str = (
        "Coverage is preliminary language detection and does not represent "
        "translation equivalence or final compliance."
    )


class LanguageAverageConfidenceResponse(ApiSchema):
    id: float | None = Field(default=None, ge=0, le=1)
    en: float | None = Field(default=None, ge=0, le=1)
    zh: float | None = Field(default=None, ge=0, le=1)


class LanguageSummaryResponse(ApiSchema):
    run_id: UUID
    total_blocks: int = Field(ge=0)
    eligible_blocks: int = Field(ge=0)
    detected_blocks: int = Field(ge=0)
    unknown_blocks: int = Field(ge=0)
    mixed_blocks: int = Field(ge=0)
    indonesian_blocks: int = Field(ge=0)
    english_blocks: int = Field(ge=0)
    chinese_blocks: int = Field(ge=0)
    other_blocks: int = Field(ge=0)
    total_characters: int = Field(ge=0)
    indonesian_characters: int = Field(ge=0)
    english_characters: int = Field(ge=0)
    chinese_characters: int = Field(ge=0)
    mixed_characters: int = Field(ge=0)
    unknown_characters: int = Field(ge=0)
    average_confidence: float | None = Field(default=None, ge=0, le=1)
    average_confidence_by_language: LanguageAverageConfidenceResponse
    language_presence: LanguagePresenceResponse
    coverage: LanguageCoverageResponse
    preliminary_label: str = "Preliminary Coverage"


class LanguageDetectionRunResponse(LanguageSummaryResponse):
    document_file_id: UUID
    document_id: UUID
    document_revision_id: UUID
    extraction_run_id: UUID
    ocr_run_id: UUID | None = None
    job_id: UUID
    detector_name: str
    detector_version: str
    status: LanguageDetectionRunStatus
    source_content_hash: str
    warnings: list[str] = Field(default_factory=list)
    metadata: dict[str, object] | None = None
    requested_by: LanguageJobRequesterReference | None = None
    started_at: datetime
    completed_at: datetime
    created_at: datetime
    is_latest: bool


class LanguageDetectionHistoryItem(ApiSchema):
    id: UUID
    job_id: UUID
    detector_name: str
    detector_version: str
    status: LanguageDetectionRunStatus
    source_content_hash: str
    total_blocks: int = Field(ge=0)
    detected_blocks: int = Field(ge=0)
    unknown_blocks: int = Field(ge=0)
    average_confidence: float | None = Field(default=None, ge=0, le=1)
    requested_by: LanguageJobRequesterReference | None = None
    redetection_reason: str | None = None
    completed_at: datetime
    is_latest: bool


class LanguageBlockResultResponse(ApiSchema):
    id: UUID
    extracted_block_id: UUID | None = None
    ocr_block_id: UUID | None = None
    container_id: UUID | None = None
    source_type: LanguageSourceType
    source_reference: str
    text: str
    language_code: LanguageCode
    primary_language_code: LanguageCode
    confidence: float = Field(ge=0, le=1)
    is_mixed: bool
    detected_languages: list[dict[str, object]]
    script_statistics: dict[str, object]
    eligibility_status: LanguageEligibilityStatus
    eligibility_reason: LanguageEligibilityReason | None = None
    character_count: int = Field(ge=0)
    latin_character_count: int = Field(ge=0)
    han_character_count: int = Field(ge=0)
    word_count: int = Field(ge=0)
    source_confidence: float | None = Field(default=None, ge=0, le=1)
    metadata: dict[str, object] | None = None
    created_at: datetime


class LanguageBlockResultListResponse(PaginationData[LanguageBlockResultResponse]):
    pass


class LanguageContainerSummaryResponse(ApiSchema):
    id: UUID
    container_id: UUID | None = None
    container_type: str
    container_name: str | None = None
    container_index: int = Field(ge=0)
    total_blocks: int = Field(ge=0)
    eligible_blocks: int = Field(ge=0)
    indonesian_blocks: int = Field(ge=0)
    english_blocks: int = Field(ge=0)
    chinese_blocks: int = Field(ge=0)
    mixed_blocks: int = Field(ge=0)
    unknown_blocks: int = Field(ge=0)
    other_blocks: int = Field(ge=0)
    indonesian_characters: int = Field(ge=0)
    english_characters: int = Field(ge=0)
    chinese_characters: int = Field(ge=0)
    mixed_characters: int = Field(ge=0)
    unknown_characters: int = Field(ge=0)
    dominant_language: LanguageCode
    language_presence: dict[str, object]
    coverage: dict[str, object]
    created_at: datetime


class LanguageContainerSummaryListResponse(
    PaginationData[LanguageContainerSummaryResponse]
):
    pass


LanguageExportFormat = Literal["json", "xlsx"]

File: frontend/src/types/languageDetection.ts
import type { DocumentUserSummary } from './document';
import type { PaginatedData, SortOrder } from './masterData';
import type {
  ExtractionDocumentSummary,
  ExtractionFileSummary,
  ExtractionJobStatus,
  ExtractionRevisionSummary,
} from './extraction';
import type { OCRJobStatus } from './ocr';

export const languageCodes = ['id', 'en', 'zh', 'mixed', 'unknown', 'other'] as const;

export type LanguageCode = (typeof languageCodes)[number];

export const languageDetectionJobStatuses = [
  'QUEUED',
  'LOADING_CONTENT',
  'DETECTING',
  'AGGREGATING',
  'PERSISTING',
  'COMPLETED',
  'PARTIALLY_COMPLETED',
  'FAILED',
  'CANCEL_REQUESTED',
  'CANCELLED',
] as const;

export type LanguageDetectionJobStatus = (typeof languageDetectionJobStatuses)[number];
export type LanguageDetectionDocumentStatus =
  'NOT_STARTED' | LanguageDetectionJobStatus;

export const terminalLanguageDetectionStatuses = [
  'COMPLETED',
  'PARTIALLY_COMPLETED',
  'FAILED',
  'CANCELLED',
] as const satisfies readonly LanguageDetectionJobStatus[];

export const activeLanguageDetectionStatuses = [
  'QUEUED',
  'LOADING_CONTENT',
  'DETECTING',
  'AGGREGATING',
  'PERSISTING',
  'CANCEL_REQUESTED',
] as const satisfies readonly LanguageDetectionJobStatus[];

export type LanguageDetectionRunStatus = 'COMPLETED' | 'PARTIALLY_COMPLETED' | 'FAILED';
export type LanguageSourceType = 'NATIVE_EXTRACTION' | 'OCR';
export type LanguageEligibilityStatus = 'ELIGIBLE' | 'INELIGIBLE';
export type LanguagePresenceStatus =
  'PRESENT' | 'NOT_PRESENT' | 'INSUFFICIENT_EVIDENCE';

export interface LanguageDetectionStartRequest {
  documentFileId: string;
  extractionRunId: string;
  ocrRunId: string | null;
  force: boolean;
}

export interface LanguageRedetectRequest {
  reason: string;
}

export interface LanguageDetectionQueuedResult {
  jobId: string;
  status: LanguageDetectionJobStatus;
  progress: number;
  documentFileId: string;
  extractionRunId: string;
  ocrRunId: string | null;
  runId: string | null;
  reusedExistingResult: boolean;
}

export interface LanguageDetectionError {
  code: string;
  message: string;
}

export interface LanguageDetectionJobListItem {
  id: string;
  document: ExtractionDocumentSummary;
  revision: ExtractionRevisionSummary;
  file: ExtractionFileSummary;
  extractionRunId: string;
  ocrRunId: string | null;
  jobType: 'INITIAL_DETECTION' | 'RE_DETECTION';
  status: LanguageDetectionJobStatus;
  progress: number;
  currentStage: string | null;
  requestedBy: DocumentUserSummary | null;
  requestedAt: string;
  startedAt: string | null;
  completedAt: string | null;
  cancelledAt: string | null;
  runId: string | null;
  resultSummary: Record<string, unknown> | null;
}

export interface LanguageDetectionJob extends LanguageDetectionJobListItem {
  attemptNumber: number;
  maximumAttempts: number;
  failedAt: string | null;
  error: LanguageDetectionError | null;
}

export interface LanguageDetectionJobListParams {
  search?: string;
  departmentId?: string;
  documentId?: string;
  revisionId?: string;
  documentFileId?: string;
  status?: LanguageDetectionJobStatus | readonly LanguageDetectionJobStatus[];
  requestedBy?: string;
  requestedFrom?: string;
  requestedTo?: string;
  page: number;
  pageSize: number;
  sortBy?: 'requestedAt' | 'completedAt' | 'status' | 'progress';
  sortOrder?: SortOrder;
}

export type LanguageDetectionJobList = PaginatedData<LanguageDetectionJobListItem>;

export interface LanguageDetectionDocumentListItem {
  document: ExtractionDocumentSummary;
  revision: ExtractionRevisionSummary;
  file: ExtractionFileSummary;
  extractionStatus: ExtractionJobStatus | null;
  ocrStatus: OCRJobStatus | null;
  languageDetectionStatus: LanguageDetectionJobStatus | null;
  languageProgress: number | null;
  languageCurrentStage: string | null;
  extractionRunId: string | null;
  ocrRunId: string | null;
  languageDetectionRunId: string | null;
  languagePresence: LanguagePresenceSummary | null;
  lastDetected: string | null;
  sourceReady: boolean;
}

export interface LanguageDetectionDocumentListParams {
  search?: string;
  departmentId?: string;
  status?: LanguageDetectionDocumentStatus;
  page: number;
  pageSize: number;
  sortBy?: 'documentCode' | 'filename' | 'uploadedAt';
  sortOrder?: SortOrder;
}

export type LanguageDetectionDocumentList =
  PaginatedData<LanguageDetectionDocumentListItem>;

export interface LanguageScore {
  languageCode: LanguageCode;
  score: number;
}

export interface ScriptStatistics {
  latinCharacterCount: number;
  hanCharacterCount: number;
  digitCount: number;
  punctuationCount: number;
  dominantScript: 'LATIN' | 'HAN' | 'MIXED' | 'NONE';
  hanRatio: number;
}

export interface LanguageCoverageValues {
  id: number;
  en: number;
  zh: number;
  mixed: number;
  unknown: number;
  other: number;
}

export interface LanguageCoverage {
  blockCoverage: LanguageCoverageValues;
  characterCoverage: LanguageCoverageValues;
  preliminary: boolean;
  disclaimer: string;
}

export interface LanguagePresenceSummary {
  id: LanguagePresenceStatus;
  en: LanguagePresenceStatus;
  zh: LanguagePresenceStatus;
}

export interface LanguageAverageConfidence {
  id: number | null;
  en: number | null;
  zh: number | null;
}

export interface LanguageSummary {
  runId: string;
  totalBlocks: number;
  eligibleBlocks: number;
  detectedBlocks: number;
  unknownBlocks: number;
  mixedBlocks: number;
  indonesianBlocks: number;
  englishBlocks: number;
  chineseBlocks: number;
  otherBlocks: number;
  totalCharacters: number;
  indonesianCharacters: number;
  englishCharacters: number;
  chineseCharacters: number;
  mixedCharacters: number;
  unknownCharacters: number;
  averageConfidence: number | null;
  averageConfidenceByLanguage: LanguageAverageConfidence;
  languagePresence: LanguagePresenceSummary;
  coverage: LanguageCoverage;
  preliminaryLabel: string;
}

export interface LanguageDetectionRun extends LanguageSummary {
  documentFileId: string;
  documentId: string;
  documentRevisionId: string;
  extractionRunId: string;
  ocrRunId: string | null;
  jobId: string;
  detectorName: string;
  detectorVersion: string;
  status: LanguageDetectionRunStatus;
  sourceContentHash: string;
  warnings: string[];
  metadata: Record<string, unknown> | null;
  requestedBy: DocumentUserSummary | null;
  startedAt: string;
  completedAt: string;
  createdAt: string;
  isLatest: boolean;
}

export interface LanguageDetectionHistoryItem {
  id: string;
  jobId: string;
  detectorName: string;
  detectorVersion: string;
  status: LanguageDetectionRunStatus;
  sourceContentHash: string;
  totalBlocks: number;
  detectedBlocks: number;
  unknownBlocks: number;
  averageConfidence: number | null;
  requestedBy: DocumentUserSummary | null;
  redetectionReason: string | null;
  completedAt: string;
  isLatest: boolean;
}

export interface LanguageBlockResult {
  id: string;
  extractedBlockId: string | null;
  ocrBlockId: string | null;
  containerId: string | null;
  sourceType: LanguageSourceType;
  sourceReference: string;
  text: string;
  languageCode: LanguageCode;
  primaryLanguageCode: LanguageCode;
  confidence: number;
  sourceConfidence: number | null;
  isMixed: boolean;
  detectedLanguages: LanguageScore[];
  scriptStatistics: ScriptStatistics;
  eligibilityStatus: LanguageEligibilityStatus;
  eligibilityReason: string | null;
  characterCount: number;
  latinCharacterCount: number;
  hanCharacterCount: number;
  wordCount: number;
  metadata: Record<string, unknown> | null;
  createdAt: string;
}

export interface LanguageContainerSummary {
  id: string;
  containerId: string | null;
  containerType: string;
  containerName: string | null;
  containerIndex: number;
  totalBlocks: number;
  eligibleBlocks: number;
  indonesianBlocks: number;
  englishBlocks: number;
  chineseBlocks: number;
  mixedBlocks: number;
  unknownBlocks: number;
  otherBlocks: number;
  indonesianCharacters: number;
  englishCharacters: number;
  chineseCharacters: number;
  mixedCharacters: number;
  unknownCharacters: number;
  dominantLanguage: LanguageCode;
  languagePresence: Record<'id' | 'en' | 'zh', boolean>;
  coverage: LanguageCoverage;
  createdAt: string;
}

export interface LanguageBlockListParams {
  languageCode?: LanguageCode;
  sourceType?: LanguageSourceType;
  containerId?: string;
  minimumConfidence?: number;
  maximumConfidence?: number;
  isMixed?: boolean;
  eligibilityStatus?: LanguageEligibilityStatus;
  search?: string;
  page: number;
  pageSize: number;
}

export interface LanguageContainerListParams {
  page: number;
  pageSize: number;
}

export type LanguageBlockList = PaginatedData<LanguageBlockResult>;
export type LanguageContainerList = PaginatedData<LanguageContainerSummary>;
export type LanguageDetectionHistory = PaginatedData<LanguageDetectionHistoryItem>;

export interface LanguageHistoryParams {
  page: number;
  pageSize: number;
}

export interface LanguageDetectionCancelResult {
  id: string;
  status: LanguageDetectionJobStatus;
  progress: number;
  currentStage: string | null;
  cancelledAt: string | null;
}

export interface LanguageDownload {
  blob: Blob;
  fileName: string | null;
}

export const isTerminalLanguageDetectionStatus = (
  status: LanguageDetectionJobStatus,
): boolean =>
  terminalLanguageDetectionStatuses.includes(
    status as (typeof terminalLanguageDetectionStatuses)[number],
  );

export const isActiveLanguageDetectionStatus = (
  status: LanguageDetectionJobStatus,
): boolean =>
  activeLanguageDetectionStatuses.includes(
    status as (typeof activeLanguageDetectionStatuses)[number],
  );

export const languageLabels: Record<LanguageCode, string> = {
  id: 'Bahasa Indonesia',
  en: 'English',
  zh: '中文 / Mandarin',
  mixed: 'Mixed',
  unknown: 'Unknown',
  other: 'Other',
};
